src/hmem_neuron.c

```diff
@@ -0,0 +1,192 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "ofi_hmem.h"
+#include "ofi.h"
+
+#if HAVE_NEURON
+
+#include <dlfcn.h>
+#include "nrt/nrt.h"
+#include "nrt/nrt_experimental.h"
+
+struct neuron_ops {
+	NRT_STATUS (*nrt_tensor_allocate)(nrt_tensor_placement_t tensor_placement,
+					  int logical_nc_id, size_t size,
+					  const char *name, nrt_tensor_t **tensor);
+	void (*nrt_tensor_free)(nrt_tensor_t **tensor);
+	void *(*nrt_tensor_get_va)(const nrt_tensor_t *tensor);
+};
+
+static void *neuron_handle;
+static struct neuron_ops neuron_ops;
+
+static int neuron_dl_init(void)
+{
+	neuron_handle = dlopen("libnrt.so", RTLD_NOW);
+	if (!neuron_handle) {
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Failed to dlopen libnrt.so\n");
+		return -FI_ENOSYS;
+	}
+
+	neuron_ops.nrt_tensor_allocate = dlsym(neuron_handle, "nrt_tensor_allocate");
+	if (!neuron_ops.nrt_tensor_allocate) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find nrt_tensor_allocate\n");
+		goto err;
+	}
+
+	neuron_ops.nrt_tensor_free = dlsym(neuron_handle, "nrt_tensor_free");
+	if (!neuron_ops.nrt_tensor_free) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find nrt_tensor_free\n");
+		goto err;
+	}
+
+	neuron_ops.nrt_tensor_get_va = dlsym(neuron_handle, "nrt_tensor_get_va");
+	if (!neuron_ops.nrt_tensor_get_va) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find nrt_tensor_get_va\n");
+		goto err;
+	}
+
+	return FI_SUCCESS;
+err:
+	dlclose(neuron_handle);
+	return -FI_ENODATA;
+}
+
+static int neuron_dl_close(void)
+{
+	dlclose(neuron_handle);
+	return FI_SUCCESS;
+}
+
+int neuron_copy_to_dev(uint64_t device, void *dev, const void *host, size_t size)
+{
+	/*
+	 * Applications will use the neuron runtime to allocate and mmap a
+	 * buffer; Libfabric can access this memory directly.
+	 */
+	memcpy(dev, host, size);
+	return FI_SUCCESS;
+}
+
+int neuron_copy_from_dev(uint64_t device, void *host, const void *dev, size_t size)
+{
+	FI_WARN_ONCE(&core_prov, FI_LOG_CORE,
+		     "Copies from AWS Neuron to host memory are not supported.");
+	return -FI_ENOSYS;
+}
+
+int neuron_hmem_init(void)
+{
+	int ret;
+
+	ret = neuron_dl_init();
```

Comment:
so it looks like for neuron, we will always use the dlopen, thus we will node need an extra option `--enable-neuron-dlopen`. I wonder is there a reason for it?

Reply:
I was asked to only provide dlopen option as the neuron library install location is not going to be clear to users. The `--with-neuron` option will check for dlopen.